pkg/security/utils/rate_limiter.go

```diff
@@ -0,0 +1,380 @@
+		groupStats:   make(map[string]RateLimiterStat),
+		statsdClient: client,
+		opts:         opts,
+	}
+}
+
+// SetGroupLimiters sets of rate limiters
+func (rl *RateLimiter) SetGroupLimiters(group string, ids []string) {
+	rl.Lock()
+	defer rl.Unlock()
+
+	newLimiters := make(map[string]*Limiter)
+	for _, id := range ids {
+		if limiter, found := rl.limiters[group][id]; found {
+			newLimiters[id] = limiter
+		} else {
+			limit := defaultLimit
+			burst := defaultBurst
+
+			if l, exists := rl.opts.Limits[group][id]; exists {
+				limit = rate.Limit(l.Limit)
+				burst = l.Burst
+			}
+			newLimiters[id] = NewLimiter(limit, burst)
+		}
+	}
+	rl.limiters[group] = newLimiters
+}
+
+// AddNewLimiter adds a new rate limiter with specified limit and burst
+func (rl *RateLimiter) AddNewLimiter(group string, id string, limit rate.Limit, burst int) error {
+	rl.RLock()
+	defer rl.RUnlock()
+
+	if burst < 0 {
+		return fmt.Errorf("EINVAL")
+	}
+
+	_, ok := rl.limiters[group]
+	if !ok {
+		rl.limiters[group] = make(map[string]*Limiter)
+	}
+
+	_, ok = rl.limiters[group][id]
+	if ok {
+		return fmt.Errorf("EEXIST")
+	}
+
+	rl.limiters[group][id] = NewLimiter(limit, burst)
+
+	return nil
+}
+
+// RemoveLimiter remove a specified rate limiter
+func (rl *RateLimiter) RemoveLimiter(group string, id string) error {
+	rl.RLock()
+	defer rl.RUnlock()
+
+	_, ok := rl.limiters[group][id]
+	if !ok {
+		return fmt.Errorf("ENOENT")
+	}
+
+	delete(rl.limiters[group], id)
+
+	return nil
+}
+
+// Allow returns true if a specific consumer shall be allowed take a token
+func (rl *RateLimiter) Allow(group string, id string) bool {
+	rl.RLock()
+	defer rl.RUnlock()
+
+	limiter, ok := rl.limiters[group][id]
+	if !ok {
+		return false
+	}
+	gStats := rl.groupStats[group]
+	if limiter.Limiter.Allow() {
+		limiter.allowed++
+		gStats.Allowed++
+		rl.groupStats[group] = gStats
+		return true
+	}
+	limiter.dropped++
+	gStats.Dropped++
+	rl.groupStats[group] = gStats
+	return false
+}
+
+// UpdateLimit update the limit of a given rate limiter
+func (rl *RateLimiter) UpdateLimit(group string, id string, newLimit rate.Limit, newBurst int) error {
+	rl.RLock()
+	defer rl.RUnlock()
+
+	if newBurst < 0 {
+		return fmt.Errorf("EINVAL")
+	}
+
+	limiter, ok := rl.limiters[group][id]
+	if !ok {
+		return fmt.Errorf("ENOENT")
+	}
+
+	limiter.Limiter.SetLimit(newLimit)
+	limiter.Limiter.SetBurst(newBurst)
+
+	return nil
+}
+
+// UpdateGroupLimit update the limit of a given rate limiter
+func (rl *RateLimiter) UpdateGroupLimit(group string, newLimit rate.Limit, newBurst int) error {
+	rl.RLock()
+	defer rl.RUnlock()
+
+	if newBurst < 0 {
+		return fmt.Errorf("EINVAL")
+	}
+
+	limiters, ok := rl.limiters[group]
+	if !ok {
+		return fmt.Errorf("ENOENT")
+	}
+
+	for _, limiter := range limiters {
+		limiter.Limiter.SetLimit(newLimit)
+		limiter.Limiter.SetBurst(newBurst)
+	}
+
+	return nil
+}
+
+// GetLimiterStats gives you the current stats of a given rate limiter, and allow you to reset them
```

Comment:
Calling function "GetX" that allows BOTH getting and setting is confusing, though I get that resetting to 0 may be simple enough to not warrant a whole separate func

Reply:
Yeah, it also ensure that the operation of getting and reseting values is atomic (so it have to be done on the same func).
Would you suggest another name?

Reply:
Maybe "PopLimiterStats", cuz when you "pop" something off a stack, it suggests that you both get the value AND remove it. By resetting the limiter to 0, that's a kind of removal.

Reply:
hmm wait, the meaning of pop can be language dependent. Verbose, but "GetAndZeroLimiterStats" allows for no confusion

Reply:
Yeah it could be tricky to have a func name that reflect 100% what it does.. Thanks for trying! We'll leave as it for now